REGRESSION(289119@main)[WPE]: Build broken on ARM 32-bits

clopez opened the pull request:
#### 3ffbaf9c98fd80340cd7a2c771f9cb3071954d93
<pre>
REGRESSION(289119@main)[WPE]: Build broken on ARM 32-bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=286489">https://bugs.webkit.org/show_bug.cgi?id=286489</a>

Reviewed by Chris Dumez.

The SIMD optimized functions on GraphicsContextGLNEON.h
before 289119@main were receiving a reference to the
pointer and modifying both the contents pointed
by the pointer as well as the pointer itself (its address)

To recover back that behaviour we need to pass now a reference
to the std::span object instead of the raw pointer.

* Source/WebCore/platform/graphics/FormatConverter.cpp:
(WebCore::uint8_t&gt;):
* Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
(WebCore::SIMD::unpackOneRowOfRGBA16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfRGB16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfARGB16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfBGRA16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfRGBA4444ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort4444):
(WebCore::SIMD::unpackOneRowOfRGBA5551ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort5551):
(WebCore::SIMD::unpackOneRowOfRGB565ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort565):

Canonical link: <a href="https://commits.webkit.org/289386@main">https://commits.webkit.org/289386@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f782027e900ffbab0e0f2db8a7767f0192c5478

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86714 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6221 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91559 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14279 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67046 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24826 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78510 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4731 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32846 "Found 3 new test failures: http/tests/security/cookie-module-propagate.html http/tests/security/cookie-module.html http/tests/xmlhttprequest/cross-origin-cookie-storage.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36564 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75249 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33731 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93454 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13867 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10056 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75841 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14068 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75030 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19347 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17751 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6652 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13486 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13890 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19150 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13628 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17073 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15413 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
```diff
@@ -33,94 +33,94 @@ namespace WebCore {
     unsigned tailComponents = componentsPerRow % 16;
     unsigned componentsSize = componentsPerRow - tailComponents;
-    const uint8_t* src = reinterpret_cast<const uint8_t*>(source);
+    const uint8_t* src = reinterpret_cast<const uint8_t*>(&source[0]);
```
Could just use `source.data()` which wouldn't crash if `source` was empty.

cdumez commented on Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
```diff
@@ -133,24 +133,24 @@ ALWAYS_INLINE void unpackOneRowOfRGBA4444ToRGBA8(const uint16_t*& source, uint8_
     unsigned componentsSize = componentsPerRow - tailComponents;
 
-    uint8_t* dst = reinterpret_cast<uint8_t*>(destination);
+    uint8_t* dst = reinterpret_cast<uint8_t*>(&destination[0]);
```
`destination.data()`

cdumez commented on Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
```diff
@@ -193,26 +193,26 @@ ALWAYS_INLINE void unpackOneRowOfRGBA5551ToRGBA8(const uint16_t*& source, uint8_
     unsigned componentsSize = componentsPerRow - tailComponents;
 
-    uint8_t* dst = reinterpret_cast<uint8_t*>(destination);
+    uint8_t* dst = reinterpret_cast<uint8_t*>(&destination[0]);
```
`destination.data()`

cdumez commented on Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
```diff
@@ -257,25 +257,25 @@ ALWAYS_INLINE void unpackOneRowOfRGB565ToRGBA8(const uint16_t*& source, uint8_t*
     unsigned tailComponents = componentsPerRow % 32;
     unsigned componentsSize = componentsPerRow - tailComponents;
-    uint8_t* dst = reinterpret_cast<uint8_t*>(destination);
+    uint8_t* dst = reinterpret_cast<uint8_t*>(&destination[0]);
```
`destination.data()`

cdumez reviewed (approved):
LGTM, thanks for fixing!

clopez committed:
REGRESSION(289119@main)[WPE]: Build broken on ARM 32-bits
https://bugs.webkit.org/show_bug.cgi?id=286489

Reviewed by Chris Dumez.

The SIMD optimized functions on GraphicsContextGLNEON.h
before 289119@main were receiving a reference to the
pointer and modifying both the contents pointed
by the pointer as well as the pointer itself (its address)

To recover back that behaviour we need to pass now a reference
to the std::span object instead of the raw pointer.

* Source/WebCore/platform/graphics/FormatConverter.cpp:
(WebCore::uint8_t>):
* Source/WebCore/platform/graphics/cpu/arm/GraphicsContextGLNEON.h:
(WebCore::SIMD::unpackOneRowOfRGBA16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfRGB16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfARGB16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfBGRA16LittleToRGBA8):
(WebCore::SIMD::unpackOneRowOfRGBA4444ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort4444):
(WebCore::SIMD::unpackOneRowOfRGBA5551ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort5551):
(WebCore::SIMD::unpackOneRowOfRGB565ToRGBA8):
(WebCore::SIMD::packOneRowOfRGBA8ToUnsignedShort565):

Canonical link: https://commits.webkit.org/289386@main
